src/registry.c

```diff
@@ -0,0 +1,1196 @@
+
+        multiple = xmlGetProp(group, (const xmlChar*)"allowMultipleSelection");
+        if (multiple && xmlStrEqual(multiple, (const xmlChar*)"true"))
+            g->allow_multiple = true;
+        xmlFree(multiple);
+
+        list_init(&g->options);
+        list_append(&ctx->option_groups, &g->base.link);
+    } else {
+        free(name);
+        free(description);
+        free(brief);
+        free(vendor);
+    }
+
+    for (node = group->children; node; node = node->next) {
+        if (is_node(node, "option"))
+            parse_option(ctx, g, node, popularity);
+    }
+}
+
+static void
+parse_option_list(struct rxkb_context *ctx, xmlNode *option_list,
+                  enum rxkb_popularity popularity)
+{
+    xmlNode *node = NULL;
+
+    for (node = option_list->children; node; node = node->next) {
+        if (is_node(node, "group"))
+            parse_group(ctx, node, popularity);
+    }
+}
+
+static void
+parse_rules_xml(struct rxkb_context *ctx, xmlNode *root,
+                enum rxkb_popularity popularity)
+{
+    xmlNode *node = NULL;
+
+    for (node = root->children; node; node = node->next) {
+        if (is_node(node, "modelList"))
+            parse_model_list(ctx, node, popularity);
+        else if (is_node(node, "layoutList"))
+            parse_layout_list(ctx, node, popularity);
+        else if (is_node(node, "optionList"))
+            parse_option_list(ctx, node, popularity);
+    }
+}
+
+static void
+ATTR_PRINTF(2, 0)
+xml_error_func(void *ctx, const char *msg, ...)
+{
+    static char buf[PATH_MAX];
+    static int slen = 0;
+    va_list args;
+    int rc;
+
+    /* libxml2 prints IO errors from bad includes paths by
+     * calling the error function once per word. So we get to
+     * re-assemble the message here and print it when we get
+     * the line break. My enthusiasm about this is indescribable.
+     */
+    va_start(args, msg);
+    rc = vsnprintf(&buf[slen], sizeof(buf) - slen, msg, args);
+    va_end(args);
+
+    /* This shouldn't really happen */
+    if (rc < 0) {
+        log_err(ctx, "+++ out of cheese error. redo from start +++\n");
+        slen = 0;
+        memset(buf, 0, sizeof(buf));
+        return;
+    }
+
+    slen += rc;
+    if (slen >= (int)sizeof(buf)) {
+        /* truncated, let's flush this */
+        buf[sizeof(buf) - 1] = '\n';
+        slen = sizeof(buf);
+    }
+
+    /* We're assuming here that the last character is \n. */
+    if (buf[slen - 1] == '\n') {
+        log_err(ctx, "%s", buf);
+        memset(buf, 0, sizeof(buf));
+        slen = 0;
+    }
+}
+
+static bool
+validate(struct rxkb_context *ctx, xmlDoc *doc)
+{
+    bool success = false;
+    xmlValidCtxt *dtdvalid = NULL;
+    xmlDtd *dtd = NULL;
+    xmlParserInputBufferPtr buf = NULL;
+    /* This is a modified version of the xkeyboard-config xkb.dtd, we
+     * allow for more either of modelList, layoutList or optionList to be
+     * missing.
+     */
+    const char dtdstr[] =
```

Comment:
My C-fu is a little weak -- will it put this in rodata or on the stack? I'd prefer if it was in rodata, I don't remember if that requires `const char *` + `strlen` or if `const char[]` + `sizeof` also works.

Reply:
```
objcopy build/libxkbregistry.so /dev/null --dump-section .rodata=/dev/stdout
```
lists the DTD content so looks like we're good here